### ISSUE TITLE
bad content-length default

### ISSUE THREAD

**Issue by reteptilian** [pallets/flask#1289]
In werkzeug's WSGIRequestHandler, CONTENT-LENGTH defaults to an empty string ('') if the client did not provide a Content-Length header.

This is causing me problems in my flask web service, which is trying to proxy requests.  The user creates a request without Content-Length, flask gives it to me and then I pass it along to another server, which then returns HTTP 400 due to the bad Content-Length that werkzeug has inserted.  According to RFC 2616, Content-Length must be a number.

Initially I thought this should be considered a werkzeug bug but maybe the empty string was intended to be werkzeug's signifier for a non-existent Content-Length, rather than an actual Content-Length value that could be used directly.  If that's the case, then I would modify flask so that Content-Length isn't included in the incoming request's headers if it is the empty string.

Bottom line, I believe that a client request that is missing a Content-Length header should not result in an incoming flask request with the Content-Length equal to the empty string.

**Comment by keyan** [comment]
The `beginner friendly` tag might be too generous, but I'm trying to tag some stuff besides documentation and this seems like a reasonably self contained fix. That is, if we can get to a consensus on whether we actually want to change Werkzeug's behavior as you suggested @untitaker.

**Comment by KWinston** [comment]
Is it the way you are creating the response? I have tested it by generating a response header without a Content-Length and then running it through flask which routes it through nginx.

Result: When you throw the response just in flask its fine (follows PEP-333).
But when you try to send it through nginx it'll crash because it does not like what is passed, the empty content-length (follows RFC 2616). For compatibility, I would agree with @untitaker that Flask omit empty Content-Type and or Content-Length. It also states in RFC 2616 that services should take priority.

**Comment by untitaker** [comment]
We should let this @mitsuhiko decide... maybe Werkzeug's current behavior is also due to other software's behavior.

**Comment by omarkhd** [comment]
And what did he decide? I'd be glad to start contributing with some code.

**Comment by untitaker** [comment]
We did not yet contact him.

I'm glad that you want to help @omarkhd -- feel free to ping mitsuhiko on the #pocoo IRC channel!

**Comment by KWinston** [comment]
For anyone seeking a quick fix solution to this error, take a look at the Proposed Solution section under Issue #1289.
[Link to Proposed Solution](https://winstonkouch.wordpress.com/2015/04/06/attempt-to-fix-some-issues-in-flask/#this-is-problem)

**Comment by untitaker** [comment]
I'm fine with the proposed solution, but please let @mitsuhiko confirm it, on the unlikely off-chance that this behavior is actually intended.

**Comment by matheuspoleza** [comment]
@KWinston can I take this solution? Its the better way to solve the problem?

**Comment by KWinston** [comment]
@matheuspoleza yes, you can use the solution I provided. The proposed solution should stop your application from crashing if you are having issues with CONTENT-LENGTH not being a number.

**Comment by matheuspoleza** [comment]
@KWinston ok, I'll do that

**Comment by ferbecab03** [comment]
@mitsuhiko : Is the proposed solution ok to implement?
@KWinston  : May I use your solution?

**Comment by mitsuhiko** [comment]
I do not understand the proposed solution. Where do we generate an empty content length string?

**Comment by ferbecab03** [comment]
> In werkzeug's WSGIRequestHandler, CONTENT-LENGTH defaults to an empty string ('') if the client did not provide a Content-Length header.<

When used with other applications that default to an empty string or a non-number passed through flask. It is more a compatibility issue with other apps.

Probably more of a documentation addition to incorporate this when using other apps while running flask?

**Comment by mitsuhiko** [comment]
I'm not sure I follow. Surely an empty string is a valid value for `CONTENT_LENGTH` so should this not be reported against whatever application fails?

**Comment by mitsuhiko** [comment]
Can someone make a minimal test case of what is wrong and what is expected? I'm at a loss here. `WSGIRequestHandler` handles incoming data only. So that should be irrelevant here, no?

**Comment by reteptilian** [comment]
@mitsuhiko according to my reading of rfc2616, you can either (1) have a content-length header, in which case it must be a number, or 2) if the header is not provided, it explains how to compute it. The situation here has the client does not provide the header ... but then flask/werkzeug artificially creates a header that is the empty string, which is not legal according to rfc 2616. I am traveling this weekend but can make a test case next week.

**Comment by untitaker** [comment]
The HTTP RFC is irrelevant here. An empty WSGI environ value is to be treated as nonexistent according to the WSGI PEP.

**Comment by untitaker** [comment]
However, if you find an abstraction over the WSGI env (within Flask/Werkzeug or outside) whose behavior differs when given an empty string instead of a nonexistent value, it's totally a bug.

**Comment by mitsuhiko** [comment]
@reteptilian as @untitaker mentioned the rfc does not matter here because this is an internal header on the incoming side and we hide it. This really needs a clear testcase.

**Comment by reteptilian** [comment]
echoheaders.py:

    import flask
    app = flask.Flask(__name__)

    @app.route('/')
    def echo_headers():
        return flask.jsonify(dict(flask.request.headers))

Run it:

    $ FLASK_APP=echoheaders python -m flask run
     * Serving Flask app "echoheaders"
     * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

Test:

    $ curl http://localhost:5000
    {
      "Accept": "*/*",
      "Content-Length": "",    <-----------   THIS IS MY COMPLAINT  ------------
      "Content-Type": "",
      "Host": "localhost:5000",
      "User-Agent": "curl/7.19.7 (x86_64-redhat-linux-gnu) libcurl/7.19.7 NSS/3.19.1 Basic ECC zlib/1.2.3 libidn/1.18 libssh2/1.4.2"
    }

**The client did not provide a Content-Length header, but flask creates one that is set to an illegal empty string.**  

_This is problematic for us, because my team made the (questionable) decision to use flask as a kind of smart proxy service, so it forwards all the client headers along to the destination, which then rightly complains (per RFC 2616) that it cannot parse the value from the Content-Length header_.  So we had to write our own code to overwrite flask's version with a correct Content-Length.  It would be nicer if flask did not artificially create this header, or at the very least, set it to a valid value (zero in this case).

**Comment by untitaker** [comment]
That is clearer. Your previous comments were about the WSGI Value, not the header abstraction.

**Comment by mitsuhiko** [comment]
I see. So unfortunately this is something that's not actually up to Flask. You would have to file this against Werkzeug. However there is a limit to what we can do with the `EnvironHeaders` (which is what drives this).

I could imagine skipping over empty content-length and content-type headers however then you can't tell a client actually sending an empty header apart. This is because this is really a limitation in the underlying WSGI protocol.

**Comment by mitsuhiko** [comment]
Closing this for now, because this is definitely a [Werkzeug](https://github.com/pallets/werkzeug) issue.

**Comment by reteptilian** [comment]
@untitaker Do you agree that I should open it against Werkzeug?  It sounded to me like you were arguing this was a flask issue not werkzeug.

My apologies for the confusion, I guess I had debugged it down to werkzeug and was trying to be helpful about the root cause but should have stuck with the simple flask example at the start.

**Comment by untitaker** [comment]
Yes it is a Werkzeug bug!

**Comment by reteptilian** [comment]
OK, all set.  

PS, thanks for the great work on these projects.  They really are fantastic.
